### PR TITLE
Add considerations for extension contexts.

### LINKED ISSUE #47: `Sec-Fetch-Site` for requests from extension background pages
(This is partly related to https://github.com/w3c/webappsec-fetch-metadata/issues/12 and was spurred by the discussion in https://chromium-review.googlesource.com/c/chromium/src/+/1903362)

Many browsers have extensions with a notion of a background page which executes scripts in its own origin, e.g. `chrome-extension://...`. These pages can load web content, in which case requests sent to the server hosting the resource need to attach `Sec-Fetch-*` headers. 

The question is what `site` value such requests should carry, with the complication being that some extensions may be granted explicit privileges to a given origin, so an extension may make both privileged and unprivileged requests. I think there are three options:

1. **Don't distinguish** privileged and unprivileged requests and always send the same `site` value, e.g. `none`.

2. Distinguish privileged and unprivileged requests and default to **treating requests from extensions as more trusted**:
- Send `none` for unprivileged requests.
- Send `same-origin` for privileged requests.

3. Distinguish privileged and unprivileged requests and default to **treating requests from extensions as less trusted**:
- Send `cross-site` for unprivileged requests.
- Send `none` for privileged requests.

My preference is (2) because (3) places a burden on web developers: they will see `cross-site` requests to their applications from users' extensions and -- in the absence of other information -- they will not be able to understand where the requests are coming from. Consequently, they will not be able to enforce security restrictions based on Fetch Metadata because blocking `cross-site` requests will appear to break their applications.

Browsers that provide information about requests sent from extensions (e.g. setting the `Origin` header to a value which identifies the extension) will still allow developers to reject such requests if they wish to do so;  given that malicious extensions are generally not in web developers' threat model and that there are significant deployability benefits of treating their requests as more trusted, I'd propose going with (2).

@mikewest @anforowicz @jerryzz0 @lweichselbaum

### PR DESCRIPTION
**Pull request by mikewest**
Mostly swiped this text from @iVanlIsh and @arturjanc's conversation in #47.

Closes #47 and #12.

### CODE REVIEW

**Comment by mikewest** [comment]
@arturjanc @iVanlIsh @annevk: Would y'all mind taking a look at this?

File: index.bs [hunk 1]
```diff
@@ -389,7 +389,29 @@ would be ideal to align on behavior for those which are likely to be common. Sme
...
+    server's Fetch Metadata logic if it chooses to do so. This increases developers' ability to
+    confidently deploy Fetch Metadata protections without undermining legitimate user interests.
+    
+With these goals in mind, user agents are encouraged to implement the following behaviors:
+
+1.  If an extension does not have permission to access to a given URL, its requests to that URL should
+    contain a `Sec-Fetch-Site` header whose value is `cross-site`, just as a normal web request would.
+    If the extension does have access to a given URL, the `Sec-Fetch-Site` value should be `same-origin`.
+
+2.  The `Origin` header should be included with outgoing requests from extension contexts.
```
Comment (annevk) [inline index.bs:414]: What value should it have though?

**Review by annevk — COMMENTED** [review]
This largely seems okay, except that I don't think this should be the place to start with normative statements around extensions. Encouraging user agents to align along these lines seems okay.

**Comment by mikewest** [comment]
> This largely seems okay, except that I don't think this should be the place to start with normative statements around extensions. Encouraging user agents to align along these lines seems okay.

I was trying to walk that line with words like "it would be ideal" and "user agents are encouraged to implement". What struck you as overly prescriptive?

File: index.bs (hunk 1 above)
Comment (mikewest) [reply index.bs:414]: Something something user-agent-defined?

Comment (mikewest) [reply index.bs:414]: "The `Origin` header should be inclued with outgoing requests from extension contexts, with a user-agent-defined value that allows servers to distinguish extension-initiated requests from web-initiated requests."?

**Comment by annevk** [comment]
All the RFC 2119 terminology. See the last paragraph of https://infra.spec.whatwg.org/#conformance for replacements. 😊

File: index.bs (hunk 1 above)
Comment (annevk) [reply index.bs:414]: `<a>implementation-defined</a>` is a thing these days, but sounds good otherwise.

Comment (mikewest) [reply index.bs]: Done.

**Comment by mikewest** [comment]
I've replaced RFC2119 words with non-normative alternatives, and resisted the urge to dive into 6919. WDYT?

File: index.bs [hunk 2]
```diff
@@ -389,7 +389,31 @@ would be ideal to align on behavior for those which are likely to be common. Sme
...
+Some user agents support extensions which are capable of making requests, and which may be empowered
+beyond regular web content in order to give users more control over their web experience. While these
+are somewhat outside the scope of the web platform, user agents are encouraged to carefully consider
+how these requests should be represented to servers. In general, it would be ideal for user agents to
+satisfy two goals:
+
+1.  Extensions without special privileges for a given site cannot cause requests that would bypass
+    that site's server-side Fetch Metadata logic.
+
+2.  Developer can recognize extensions' requests so that they can be exempted from the server's Fetch
```
Comment (dergoegge) [inline index.bs:404]:
```suggestion
2.  Developers can recognize extensions' requests so that they can be exempted from the server's Fetch
```

**Review by dergoegge — APPROVED** [review]
Text looks good to me.

File: index.bs [hunk 3]
```diff
@@ -389,7 +389,31 @@ would be ideal to align on behavior for those which are likely to be common. Sme
     user agent's bookmark bar, another app entirely, etc), then `Sec-Fetch-Site: none` may be
     appropriate.
 
-
+Extension-Initiated Requests {#extension-initiated}
+---------------------------------------------------
+
+Some user agents support extensions which are capable of making requests, and which may be empowered
```
Comment (annevk) [inline index.bs:395]:
```suggestion
Some user agents support extensions which are capable of making requests, and which can be empowered
```

File: index.bs [hunk 4]
```diff
@@ -389,7 +389,31 @@ would be ideal to align on behavior for those which are likely to be common. Sme
...
+beyond regular web content in order to give users more control over their web experience. While these
+are somewhat outside the scope of the web platform, user agents are encouraged to carefully consider
+how these requests should be represented to servers. In general, it would be ideal for user agents to
```
Comment (annevk) [inline index.bs:398]:
```suggestion
how these requests are to be represented to servers. In general, it would be ideal for user agents to
```

**Review by annevk — APPROVED** [review]
Modulo nits this seems fine, thanks!

**Comment by arturjanc** [comment]
I took a quick glance at the change and it looks reasonable overall (thanks, @mikewest!)

Ideally, we would be a little more forceful and tell implementers that behaviors such as always sending the `Origin` header on extension-initiated requests are important to ensure that sites which use Fetch Metadata headers don't break extensions. But I don't know how to do this non-normatively so ¯\\_(ツ)_/¯

**Pull request merged**